gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/intercept/PubSubClientIntroductionAdvice.java

```diff
@@ -0,0 +1,173 @@
+        this.publisherFactory = publisherFactory;
+        this.serDesRegistry = serDesRegistry;
+        this.scheduler = Schedulers.from(executorService);
+        this.conversionService = conversionService;
+    }
+
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+
+        if (context.hasAnnotation(Topic.class)) {
+            Map<String, String> messageAttributes = new HashMap<>();
+            AnnotationValue<Topic> topic = context.getAnnotation(Topic.class);
+            String contentType = topic.get("contentType", String.class).orElse("");
+            List<AnnotationValue<Header>> headerAnnotations = context.getAnnotationValuesByType(Header.class);
+
+            PubSubMessageSerDes serDes = serDesRegistry.find(contentType)
+                    .orElseThrow(() -> new PubSubClientException("Could not locate a valid SerDes implementation for type: " + contentType));
+            Publisher publisher = publisherFactory.createPublisher(topic.getValue(String.class).orElse(""));
+            Argument<?> bodyArgument = findBodyArgument(context.getExecutableMethod())
+                    .orElseThrow(() -> new PubSubClientException("No valid message body argument found for method: " + context.getExecutableMethod()));
+
+            Map<String, Object> parameterValues = context.getParameterValueMap();
+
+            headerAnnotations.forEach((header) -> {
+                String name = header.get("name", String.class).orElse(null);
+                String value = header.getValue(String.class).orElse(null);
+                if (StringUtils.isNotEmpty(name) && StringUtils.isNotEmpty(value)) {
+                    messageAttributes.put(name, value);
+                }
+            });
+
+            ReturnType<Object> returnType = context.getReturnType();
+            Class<?> javaReturnType = returnType.getType();
+
+
+            Argument[] arguments = context.getArguments();
+            for (Argument arg : arguments) {
+                AnnotationValue<Header> headerAnn = arg.getAnnotation(Header.class);
+                if (headerAnn != null) {
+                    Map.Entry<String, String> entry = getNameAndValue(arg, headerAnn, parameterValues);
+                    messageAttributes.put(entry.getKey(), entry.getValue());
+                }
+            }
+            Object body = parameterValues.get(bodyArgument.getName());
+            PubsubMessage pubsubMessage = null;
+            if (body.getClass() == PubsubMessage.class) {
+                pubsubMessage = (PubsubMessage) body;
+            } else {
+                //if target type is byte[] we bypass serdes completely
+                byte[] serialized = null;
+                if (body.getClass() == byte[].class) {
+                    serialized = (byte[]) body;
+                    messageAttributes.put("Content-Type", MediaType.APPLICATION_OCTET_STREAM);
```

Comment:
This seems to be ignoring the content type set at the annotation. This would prevent folks from sending bytes with other content types. Is that the desired behavior?

Reply:
You've got a good point. Content-Type handling is always a struggle. But you are right, folks could send an `application/image` and push bytes. I guess the only downside of this is that since the default content type is `application/json` someone could send `byte[]` and have `application/json` set as default. We could set `application/json` as default only for types that are not `byte[]` or `PubSubMessage` (need to handle primitives too), for everything else we would just not append or append what is in the annotation. 
Open for feedback...

Reply:
I would say to allow for `byte[]` with JSON. The documentation will outline that is the default content type and it's up to users to change it. Having special rules for different scenarios will only lead to confusion and docs that are difficult to write and read.